docs/provider-configuration.md
## cloud-provider-oci Configuration Options

This document outlines all of the configuration options available for the Oracle
Cloud Infrastructure (OCI) Kubernetes cloud-provider (cloud-provider-oci).

## Contents

- [cloud-provider-oci Configuration Options](#cloud-provider-oci-configuration-options)
- [Contents](#contents)
- [Cloud Configuration File](#cloud-configuration-file)
    - [Sample configuration](#sample-configuration)
    - [Configuration Options](#configuration-options)
        - [Top Level](#top-level)
            - [Top Level Required Parameters](#top-level-required-parameters)
            - [Top Level Optional Parameters](#top-level-optional-parameters)
        - [Auth](#auth)
            - [Auth Required Parameters](#auth-required-parameters)
            - [Auth Optional Parameters](#auth-optional-parameters)
        - [Loadbalancer](#loadbalancer)
            - [Loadbalancer Optional Parameters](#loadbalancer-optional-parameters)


## Cloud Configuration File

Kubernetes knows how to interact with OCI via the file `cloud-provider.yaml`.
It is a standard YAML file that provides Kubernetes with user authentication
credentials and additional configuration specific to your OCI account.

### Sample configuration

This is an example of a typical configuration that touches the values that most
often need to be set. It provides details for how to authenticate with OCI via a
standard IAM user and configures load balancer provisioning.

```yaml
auth:
  region: us-ashburn-1
  tenancy: ocid1.tenancy.oc1..
  user: ocid1.user.oc1..
  key: |
    -----BEGIN RSA PRIVATE KEY-----
    <snip>
    -----END RSA PRIVATE KEY-----
  fingerprint: 8c:bf:17:7b:5f:e0:7d:13:75:11:d6:39:0d:e2:84:74

compartment: ocid1.compartment.oc1..
vcn: ocid1.vcn.oc1..

loadBalancer:
  subnet1: ocid1.subnet.oc1.iad..
  subnet2: ocid1.subnet.oc1.iad..
```

### Configuration Options

The OCI Cloud Provider offers a wide range of configuration options for each
service it supports. The currently available configuration sections include:

 - [Top Level](#top-level)
 - [Auth](#auth)
 - [Load Balancer](#load-balancer)

#### Top Level

These configuration options for the OCI cloud provider relate to its global
configuration and should appear at the top level of your `cloud-provider.yaml`
file.

##### Top Level Required Parameters

- `compartment`: The OCID of the OCI Compartment within which your Kubernetes
  cluster resides.
- `vcn`: The OCID of the Virtual Cloud Network (VCN) within which your
  Kubernetes cluster resides.

##### Top Level Optional Parameters

- `useInstancePrincipals`: `true`/`false`. If enabled use [Instance Principals][1]
  authentication for communication with OCI services.

#### Auth

These configuration options for the OCI cloud provider relate to authentication
with OCI services via an IAM user and OCI API key. For more information please
see [here][2].

**NOTE**: If `useInstancePrincipals` is set to `true` then all auth fields are
optional.

##### Auth Required Parameters

 - `region`: The identifier for the OCI Region in which your Kubernetes cluster
   resides (e.g. `us-ashburn-1`).
 - `tenancy`: The OCID of your OCI Tenancy.
 - `user`: The OCID of the IAM user as which cloud-provider OCI will
   authenticate with OCI services.
 - `key`: The PEM encoded OCI API private key associated with the `user` as
   which cloud-provider OCI will authenticate with OCI services.
 - `fingerprint`: The fingerprint of the OCI API public key.

##### Auth Optional Parameters

 - `regionKey`: The short identifier for the OCI Region in which your Kubernetes
   cluster resides (e.g. `iad`).
 - `passphrase`: The passphrase if the provided private key is encrypted.

#### Load balancer

These configuration options for the OCI cloud provider relate to provisioning
loadbalancers via the cloud controller manager.

##### Loadbalancer Optional Parameters

 - `disabled`:  `true`/`false`. Disables the creation of load balancers.
 - `securityListManagementMode`: `All` (default), `Frontend`, and `None`.
   Defines how the oci-cloud-controller-manager manages Security Lists to enable
   traffic to ingress/egress the Kubernetes cluster via loadbalancers.

   Management mode `All` enables management of all required Security List rules
   for traffic to ingress/egress the Kubernetes cluster via loadbalancers.

   Management mode `Frontend` enables management of Security List rules for
   loadbalancer ingress only.

   Management mode `None` disables Security List management by the
   oci-cloud-controller-manager completely. This requires that the user
   has setup a rule that allows inbound traffic to the appropriate ports
   for the kube proxy health check port, node port ranges, and health check port
   ranges. E.g. 10.82.0.0/16 30000-32000.
 - `subnet1`: The OCID of the first of two required Subnets to which
   loadbalancers will be attached.
 - `subnet2`: The OCID of the second of two required Subnets to which
   loadbalancers will be attached.
 - `securityLists`: A map of Subnet OCIDs to Security List OCIDs to enable
   configuration of which Security Lists the oci-cloud-controller-manager
   manages when creating ingress/egress rules to enable traffic to
   ingress/egress the Kubernetes cluster via loadbalancers.


[1]: https://docs.cloud.oracle.com/iaas/Content/Identity/Tasks/callingservicesfrominstances.htm
[2]: https://docs.cloud.oracle.com/iaas/Content/API/Concepts/apisigningkey.htm


docs/rate-limiter-configuration.md
# Rate Limiting Configuration

This file defines a list of Rate Limiting configurations properties,
supported by the `oci-cloud-controller-manager`.

These properties exist in the `yaml` configuration under `rateLimiter`
in root. For example:

```yaml
auth:
  ...
loadBalancer:
  ...
rateLimiter:
  rateLimitQPSRead: ...
  rateLimitBucketRead: ...
  rateLimitQPSWrite: ...
  rateLimitBucketWrite: ...
```

## Request read/write rate limiting properties

| Name | Description | Default |
| ---- | ----------- | ------- |
| `rateLimitQPSRead` | The maximum queries allowed per second for read requests. | 20.0 |
| `rateLimitBucketRead` | The maximum token bucket burst size for read requests. | 5.0 |
| `rateLimitQPSWrite` | The maximum queries allwoed per second for write requests. | 20.0 |
| `rateLimitBucketWrite` | The maximim token bucket burst size for write requests. | 5.0 |

## Disable Rate Limiting
The rate limiting can be completely disabled by adding a property `disableRateLimiter: true`.
By default the property is `false`

```yaml
auth:
  ...
loadBalancer:
  ...
rateLimiter:
  disableRateLimiter: true
```


CHANGELOG.md
# CHANGELOG

## 0.7.0

* Support assigning SSL certs to BackendSets [#243][30]

## 0.6.1

* Support load balancer listeners with http protocol via annotation [#239][28]
* Prevent leaking security list rules when updating a service [#238][29]

## 0.6.0

* Allow dynamic configuration of security list management via annotations. [#217][23]
* Migrate to Zap logger. [#223][24]
* Runtime checks around providerID. [#228][25]
* Support request rate limiting. [#201][26]
* Bump dependencies to Kubernetes 1.11.1. [#234][27]

## 0.5.1

* Fix panic on startup when security lists aren't managed. [#213][22]
* Update oci-go-sdk to 2.0.0 which includes support for load balancer tagging.

## 0.5.0

* Allow load balancer functionality to be disabled. [#198][20]
* Allow subnets in configuration to be optional. [#202][21]

## 0.4.0

 * Implement `loadbalancer.securityListManagementMode: Frontend` which only
   manages security list rules for load balancer ingress. [#180][16]
 * Depreciate `loadbalancer.disableSecurityListManagement` in favour of
   `loadbalancer.securityListManagementMode: None`. [#180][16]
 * Implement `loadbalancer.securityLists` to allow explicit configuration of the
   security lists that the CCM manages on a per-subnet basis [#164][17].
 * Implement support for [instance principles][19] authentication [#155][18]

## 0.3.2

 * [BUG] Fix panic when EnsureLoadBalancer() called with 0 Nodes [#176][11]
 * [BUG] Fix panic when GetInstanceByNodeName() encountered an instance without
   either a public IP or a hostname [#167][14]
 * [BUG] Fix regression where compartment OCID was no longer looked up from
   metadata when not provided in cloud-provider config [#168][15]
 * Depreciate cloud-provider config property `auth.key_passphrase` replacing it
   with `auth.passphrase` [#142][12]
 * Depreciate cloud-provider config property `auth.compartment` replacing it
   with `compartment` [#170][13]

## 0.3.1

 * Remove redundant `--cluster-cidr` flag from DaemonSet [#163][10]

## 0.3.0

 * Create load balancers using a single OCI API request significantly reducing fresh load balancer creation time [[#148][6]]
 * Fix failure to delete security list rules when deleting a Service type=LoadBalancer or changing its NodePort(s) [[#151][7]]
 * Move to ConfigMap based leader election [[#160][8]]
 * Fix potential panic when a load balancer had no backends [[#157][9]]

## 0.2.1

 * Add OCI proxy support [[#135][5]]
 * Add security rules for a services health check port [[#125][4]]

## 0.2.0

 * Wercker release process [[#131][3]]

## 0.1.2

 * Fixes a bug where existing security list rules with no destination port range
   set would cause the CCM to fail when update security lists whilst ensuring
   load balancer state [[#112][2]]

## 0.1.1

 * Fix failure to update load balancer listener port when frontend service port changed [[#92][1]]

## 0.1.0

 * Implement `cloudprovider.Instances` interface - keeps track of node state within an OCI Kubernetes cluster.
 * Implement `cloudprovider.LoadBalancer` interface - enables create, update, and delete of OCI load balancers for services of type `LoadBalancer`.
 * Implement `cloudprovider.Zones` interface - provides OCI region information for cluster nodes.

[1]: https://github.com/oracle/oci-cloud-controller-manager/issues/92
[2]: https://github.com/oracle/oci-cloud-controller-manager/issues/112
[3]: https://github.com/oracle/oci-cloud-controller-manager/issues/131
[4]: https://github.com/oracle/oci-cloud-controller-manager/issues/125
[5]: https://github.com/oracle/oci-cloud-controller-manager/issues/135
[6]: https://github.com/oracle/oci-cloud-controller-manager/issues/148
[7]: https://github.com/oracle/oci-cloud-controller-manager/issues/151
[8]: https://github.com/oracle/oci-cloud-controller-manager/issues/160
[9]: https://github.com/oracle/oci-cloud-controller-manager/issues/157
[10]: https://github.com/oracle/oci-cloud-controller-manager/issues/163
[11]: https://github.com/oracle/oci-cloud-controller-manager/issues/176
[12]: https://github.com/oracle/oci-cloud-controller-manager/issues/142
[13]: https://github.com/oracle/oci-cloud-controller-manager/issues/170
[14]: https://github.com/oracle/oci-cloud-controller-manager/issues/167
[15]: https://github.com/oracle/oci-cloud-controller-manager/issues/168
[16]: https://github.com/oracle/oci-cloud-controller-manager/issues/180
[17]: https://github.com/oracle/oci-cloud-controller-manager/issues/164
[18]: https://github.com/oracle/oci-cloud-controller-manager/issues/155
[19]: https://docs.us-phoenix-1.oraclecloud.com/Content/Identity/Tasks/callingservicesfrominstances.htm
[20]: https://github.com/oracle/oci-cloud-controller-manager/pull/199
[21]: https://github.com/oracle/oci-cloud-controller-manager/pull/204
[22]: https://github.com/oracle/oci-cloud-controller-manager/issues/213
[23]: https://github.com/oracle/oci-cloud-controller-manager/issues/217
[24]: https://github.com/oracle/oci-cloud-controller-manager/pull/223
[25]: https://github.com/oracle/oci-cloud-controller-manager/pull/228
[26]: https://github.com/oracle/oci-cloud-controller-manager/issues/108
[27]: https://github.com/oracle/oci-cloud-controller-manager/issues/232
[28]: https://github.com/oracle/oci-cloud-controller-manager/issues/208
[29]: https://github.com/oracle/oci-cloud-controller-manager/pull/238
[30]: https://github.com/oracle/oci-cloud-controller-manager/issues/235


pkg/cloudprovider/providers/oci/config/config.go
// Copyright 2017 Oracle and/or its affiliates. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package config

import (
	"context"
	"io"
	"os"

	"bitbucket.oci.oraclecorp.com/oke/oke-common/resourceprincipals"

	"github.com/oracle/oci-cloud-controller-manager/pkg/oci/instance/metadata"
	"github.com/oracle/oci-go-sdk/common"
	"github.com/oracle/oci-go-sdk/common/auth"
	"github.com/pkg/errors"
	"github.com/spf13/viper"
	"go.uber.org/zap"
	"gopkg.in/yaml.v2"
)

// AuthConfig holds the configuration required for communicating with the OCI
// API.
type AuthConfig struct {
	Region      string `yaml:"region"`
	TenancyID   string `yaml:"tenancy"`
	UserID      string `yaml:"user"`
	PrivateKey  string `yaml:"key"`
	Fingerprint string `yaml:"fingerprint"`
	Passphrase  string `yaml:"passphrase"`

	// Used by the flex driver for OCID expansion. This should be moved to top level
	// as it doesn't strictly relate to OCI authentication.
	RegionKey string `yaml:"regionKey"`

	// The fields below are deprecated and remain purely for backwards compatibility.
	// At some point these need to be removed.

	// UseInstancePrincipals is DEPRECATED should use top-level UseInstancePrincipals
	UseInstancePrincipals bool `yaml:"useInstancePrincipals"`
	// CompartmentID is DEPRECATED and should be set on the top level Config
	// struct.
	CompartmentID string `yaml:"compartment"`
	// PrivateKeyPassphrase is DEPRECATED in favour of Passphrase.
	PrivateKeyPassphrase string `yaml:"key_passphrase"`

	//Metadata service to help fill in certain fields
	metadataSvc metadata.Interface
}

const (
	// ManagementModeAll denotes the management of security list rules for load
	// balancer ingress/egress, health checkers, and worker ingress/egress.
	ManagementModeAll = "All"
	// ManagementModeFrontend denotes the management of security list rules for load
	// balancer ingress only.
	ManagementModeFrontend = "Frontend"
	// ManagementModeNone denotes the management of no security list rules.
	ManagementModeNone = "None"
)

// LoadBalancerConfig holds the configuration options for OCI load balancers.
type LoadBalancerConfig struct {
	// Disabled disables the creation of a load balancer.
	Disabled bool `yaml:"disabled"`

	// DisableSecurityListManagement disables the automatic creation of ingress
	// rules for the node subnets and egress rules for the load balancers to the node subnets.
	//
	// If security list management is disabled, then it requires that the user
	// has setup a rule that allows inbound traffic to the appropriate ports
	// for kube proxy health port, node port ranges, and health check port ranges.
	// E.g. 10.82.0.0/16 30000-32000
	DisableSecurityListManagement bool `yaml:"disableSecurityListManagement"`

	// SecurityListManagementMode defines how the CCM manages security lists
	// when provisioning load balancers. Available modes are All, Frontend,
	// and None.
	SecurityListManagementMode string `yaml:"securityListManagementMode"`

	Subnet1 string `yaml:"subnet1"`
	Subnet2 string `yaml:"subnet2"`

	// SecurityLists defines the Security List to mutate for each Subnet (
	// both load balancer and worker).
	SecurityLists map[string]string `yaml:"securityLists"`
}

// RateLimiterConfig holds the configuration options for OCI rate limiting.
type RateLimiterConfig struct {
	RateLimitQPSRead     float32 `yaml:"rateLimitQPSRead"`
	RateLimitBucketRead  int     `yaml:"rateLimitBucketRead"` //Read?
	RateLimitQPSWrite    float32 `yaml:"rateLimitQPSWrite"`
	RateLimitBucketWrite int     `yaml:"rateLimitBucketWrite"`
}

// Config holds the OCI cloud-provider config passed to Kubernetes components
// via the --cloud-config option.
type Config struct {
	Auth         AuthConfig          `yaml:"auth"`
	LoadBalancer *LoadBalancerConfig `yaml:"loadBalancer"`
	RateLimiter  *RateLimiterConfig  `yaml:"rateLimiter"`

	RegionKey string `yaml:"regionKey"`

	// When set to true, clients will use an instance principal configuration provider and ignore auth fields.
	UseInstancePrincipals bool `yaml:"useInstancePrincipals"`
	// CompartmentID is the OCID of the Compartment within which the cluster
	// resides.
	CompartmentID string `yaml:"compartment"`
	// VCNID is the OCID of the Virtual Cloud Network (VCN) within which the
	// cluster resides.
	VCNID string `yaml:"vcn"`

	// Metadata service to help fill in certain fields
	metadataSvc metadata.Interface

	// UseServicePrincipals when set to true, clients will use an instance principal to fetch the s2s token
	// from identity.
	UseServicePrincipals bool `yaml:"UseServicePrincipals"`
}

// Complete the load balancer config applying defaults / overrides.
func (c *LoadBalancerConfig) Complete() {
	if c.Disabled {
		return
	}
	if len(c.SecurityListManagementMode) == 0 {
		if c.DisableSecurityListManagement {
			zap.S().Warnf("cloud-provider config: \"loadBalancer.disableSecurityListManagement\" is DEPRECATED and will be removed in a later release. Please set \"loadBalancer.SecurityListManagementMode: %s\".", ManagementModeNone)
			c.SecurityListManagementMode = ManagementModeNone
		} else {
			c.SecurityListManagementMode = ManagementModeAll
		}
	}
}

// Complete the authentication config applying defaults / overrides.
func (c *AuthConfig) Complete() {
	if len(c.Passphrase) == 0 && len(c.PrivateKeyPassphrase) > 0 {
		zap.S().Warn("cloud-provider config: auth.key_passphrase is DEPRECIATED and will be removed in a later release. Please set auth.passphrase instead.")
		c.Passphrase = c.PrivateKeyPassphrase
	}
	if c.Region == "" || c.CompartmentID == "" {
		meta, err := c.metadataSvc.Get()
		if err != nil {
			zap.S().Warn("cloud-provider config: Unable to access metadata on instance. Will not be able to complete configuration if items are missing")
			return
		}
		if c.Region == "" {
			c.Region = meta.CanonicalRegionName
		}

		if c.CompartmentID == "" {
			c.CompartmentID = meta.CompartmentID
		}
	}
}

// Complete the top-level config applying defaults / overrides.
func (c *Config) Complete() {
	if c.LoadBalancer != nil {
		c.LoadBalancer.Complete()
	}
	c.Auth.Complete()
	// Ensure backwards compatibility fields are set correctly.
	if len(c.CompartmentID) == 0 && len(c.Auth.CompartmentID) > 0 {
		zap.S().Warn("cloud-provider config: \"auth.compartment\" is DEPRECATED and will be removed in a later release. Please set \"compartment\".")
		c.CompartmentID = c.Auth.CompartmentID
	}
	if c.Auth.UseInstancePrincipals {
		zap.S().Warn("cloud-provider config: \"auth.useInstancePrincipals\" is DEPRECATED and will be removed in a later release. Please set \"useInstancePrincipals\".")
		c.UseInstancePrincipals = true
	}

	if len(c.RegionKey) == 0 {
		if len(c.Auth.RegionKey) > 0 {
			zap.S().Warn("cloud-provider config: \"auth.RegionKey\" is DEPRECATED and will be removed in a later release. Please set \"RegionKey\".")
			c.RegionKey = c.Auth.RegionKey
		} else {
			meta, err := c.metadataSvc.Get()
			if err != nil {
				zap.S().Warn("cloud-provider config: Unable to access metadata on instance. Will not be able to complete configuration if items are missing")
				return
			}
			c.RegionKey = meta.Region
		}
	}
}

// Validate validates the OCI cloud-provider config.
func (c *Config) Validate() error {
	return ValidateConfig(c).ToAggregate()
}

// ReadConfig consumes the config Reader and constructs a Config object.
func ReadConfig(r io.Reader) (*Config, error) {
	if r == nil {
		return nil, errors.New("no cloud-provider config file given")
	}

	cfg := &Config{}
	err := yaml.NewDecoder(r).Decode(&cfg)
	if err != nil {
		return nil, errors.Wrap(err, "unmarshalling cloud-provider config")
	}

	cfg.metadataSvc = metadata.New()
	cfg.Auth.metadataSvc = cfg.metadataSvc
	// Ensure defaults are correctly set
	cfg.Complete()

	return cfg, nil
}

// FromFile will load a cloud provider configuration file from a given file path.
func FromFile(path string) (*Config, error) {
	f, err := os.Open(path)
	defer f.Close()

	if err != nil {
		return nil, err
	}

	return ReadConfig(f)
}

// NewConfigurationProvider takes a cloud provider config file and returns an OCI ConfigurationProvider
// to be consumed by the OCI SDK.
func NewConfigurationProvider(cfg *Config) (common.ConfigurationProvider, error) {

	if cfg != nil {
		err := cfg.Validate()
		if err != nil {
			return nil, errors.Wrap(err, "invalid client config")
		}
	}

	if viper.GetBool("use-resource-principal") {
		zap.S().Infof("using resource principal as the authentication mechanism")
		timeoutCtx, cancelFunc := context.WithTimeout(context.Background(), viper.GetDuration("resource-principal-initial-timeout"))
		defer cancelFunc()
		cp, err := resourceprincipals.NewConfigurationProvider(timeoutCtx, nil, common.Region(cfg.Auth.Region), viper.GetString("resource-principal-file"))
		if err != nil {
			return nil, errors.Wrap(err, "failed to instantiate ResourcePrincipalConfigurationProvider")
		}
		return cp, nil
	}

	cp, err := auth.NewServicePrincipalWithInstancePrincipalConfigurationProvider("")
	if err != nil {
		return nil, errors.Wrap(err, "failed to instantiate ServicePrincipalWithInstancePrincipalConfigurationProvider")
	}
	return cp, nil

}
